modules/apps/frontend-js/frontend-js-spa-web/test/app/App.ts

```diff
@@ -2145,102 +2383,114 @@ describe('App', function () {
 
-		this.app
-			.navigate('/path1')
+		app.navigate('/path1')
 			.then(() => {
-				return this.app.navigate('/path2');
+				return app?.navigate('/path2');
 			})
 			.then(() => {
 				expect(getUrlPathWithoutHash(document.referrer)).toBe('/path1');
 
-				return this.app.navigate('/path3');
+				return app?.navigate('/path3');
 			})
 			.then(() => {
 				expect(getUrlPathWithoutHash(document.referrer)).toBe('/path2');
-				this.app.on(
-					'endNavigate',
-					() => {
-						expect(getUrlPathWithoutHash(document.referrer)).toBe(
-							'/path1'
-						);
-						done();
-					},
-					true
-				);
+				app?.on('endNavigate', () => {
+					expect(getUrlPathWithoutHash(document.referrer)).toBe(
+						'/path1'
+					);
+					done();
+				});
 				window.history.back();
 			});
 	});
 
 	it('does reload page on navigate back to a routed page without history state and skipLoadPopstate is active', () => {
-		this.app = new App();
-		this.app.reloadPage = jest.fn();
-		this.app.addRoutes(new Route('/path1', Screen));
-		this.app.addRoutes(new Route('/path2', Screen));
+		app = new App();
+
+		app.reloadPage = jest.fn();
 
-		return this.app.navigate('/path1').then(() => {
-			window.history.replaceState(null, null, null);
+		app.addRoutes(new Route('/path1', Screen));
+		app.addRoutes(new Route('/path2', Screen));
 
-			return this.app.navigate('/path2').then(() => {
+		return app.navigate('/path1').then(() => {
+			window.history.replaceState(null, '', null);
+
+			return app?.navigate('/path2').then(() => {
 				return new Promise((resolve) => {
 					window.addEventListener(
 						'popstate',
 						() => {
-							expect(this.app.reloadPage).not.toHaveBeenCalled();
+							expect(app?.reloadPage).not.toHaveBeenCalled();
+
 							resolve();
 						},
 						{once: true}
 					);
-					this.app.skipLoadPopstate = true;
+
+					app.skipLoadPopstate = true;
 					window.history.back();
 				});
 			});
 		});
 	});
 });
 
-function enterDocumentLinkElement(href) {
+function enterDocumentLinkElement(href: string): HTMLElement {
 	document.body.appendChild(
 		buildFragment('<a id="link" href="' + href + '">link</a>')
 	);
 
-	return document.getElementById('link');
+	return document.getElementById('link')!;
 }
 
-function enterDocumentFormElement(action, method) {
+function enterDocumentFormElement(action: string, method: string): HTMLElement {
 	const random = Math.floor(Math.random() * 10000);
+
 	document.body.appendChild(
 		buildFragment(
 			`<form id="form_${random}" action="${action}" method="${method}" enctype="multipart/form-data"></form>`
 		)
 	);
 
-	return document.getElementById(`form_${random}`);
+	return document.getElementById(`form_${random}`)!;
 }
 
 function exitDocumentLinkElement() {
-	document.getElementById('link').remove();
+	document.getElementById('link')?.remove();
 }
 
-function preventDefault(event) {
+function preventDefault(event: Event) {
 	event.preventDefault();
 }
 
```

Comment:
On line 2511 (can't comment on it in this UI because it has no changes) the whole `window.location` thingy is smelly. TS throws an error `The operand of a 'delete' operator must be optional.` when we want to delete it, and also it complains about mismatching properties on the line below when we want to assign new properties to it saying `Type '{}' is missing the following properties from type 'Location': ancestorOrigins, hash, host, hostname, and 9 more.`